Animation Stage1 Alternative - animateJoin method

softwords opened the pull request:
Alternative version of Stage 1 that splits of `animateJoin`  into a separate method of `Animator`.

softwords commented:
@bluong alternative PR ...

bluong commented on src/animators/easingFunctions.ts:
```diff
@@ -0,0 +1,66 @@
+      };
+    }
+  }
+}
```
Can we extract this out (possibly into a pull request into develop if you can attach it to `Animators.Easing` or into `objectConstancy` and then we can connect the pieces later)?

softwords replied:
Not really sure what you mean by "attach to to Animators.Easing" - did you mean a namespace?
Animators.Easing is already in use by the class.

bluong replied:
Oh I mean that `Animators.Easing` can make use of these easing functions, however the signatures need to be modified to account for this.

bluong replied:
To be more specific, I think this `easingFunctions.ts` file specifically can be made as a pull request to develop, as long as `Animators.Easing`'s `easing()` endpoint is modified to make use of the above types.

softwords commented on src/drawers/drawer.ts:
```diff
@@ -106,9 +168,20 @@ export class Drawer {
+      this._joinResult.exit
+        .remove();
+      step.animator.animate(selection, step.attrToAppliedProjector);
+    }
```
good morning @bluong 
I have thought further on the issue of the change to animator interface, specifically adding and using the `animateJoin` method as in here.

First point to note is that for the majority of cases, the javascript consumer never calls the `animator` directly, it is called on their behalf by the `Drawer` when they assign the animator to their plot.

So, the only case that presents a problem is if a javascript client has built a custom animator and implemented `animate` - `animate` would not be called, and in fact there would be an error at this point because `animateJoin` is not found.

This (unattractive) change will defend against this.

I think the only circumstance now where any consumer of plottable will see a difference is that if you are authoring a custom animator in typescript, you'll get a compile warning about the missing member.

softwords commented:
@bluong could you merge develop into objectConstancy please and I'll update all the pull requests accordingly

softwords commented:
Perhaps we should now close #2921 - (I think the overloaded interface is out of consideration?)  and make this the target for future deliberation... 
Meantime I have updated the big PR #2907 with all these changes, so it will continue to represent the complete, current state of this development thread, and be the source of any future smaller pull requests.
#2907 is also the source of the fiddles http://jsfiddle.net/oaskqta0/24/  and http://jsfiddle.net/oaskqta0/27/
so we keep that verification that it all still works!

In http://jsfiddle.net/oaskqta0/27/ I have added a "legacy JS animator", that is, a class built in JS that implements the legacy animate function, to verify that this is gracefully suppported in this build:

```javascript

function myLegacyAnimator() {
    this.animate = function(selection, attrs) {
        alert("hello from the legacy JS custom animator");
        selection.attr(attrs);
    }
}
```

bluong commented:
I've updated the objectContancy branch

Commit:
[Bar Plot tests] Convert single-datum autodomaining tests

Commit:
[Bar Plot tests] Convert autoDomain() idempotence test

Commit:
[Bar Plot tests] Remove dangling test for attr("width")

The behavior there is wrong (https://github.com/palantir/plottable/issues/2536)

bluong committed:
Merge pull request #2935 from palantir/xyPlotTests

[XYPlot Tests] - Rework

Commit:
[Bar Plot tests] Adjust tests so they pass Phantom

Zero-height bars can't be selected by entitiesIn() except in Chrome
Bug filed at https://github.com/palantir/plottable/issues/2536

Commit:
[Bar Plot tests] TSLint

Commit:
[Bar Plot tests] Minor cleanup in response to CR.

Commit:
[Bar Plot tests] Move afterEach()s to occur right after related beforeEach().

Commit:
[Bar Plot tests] Switch Scale forEach() to use Scale constructors

crmorford committed:
entity nearest returns undefined instead of breaking

Commit:
Revert "[Bar Plot tests] Switch Scale forEach() to use Scale constructors"

There were some weird errors being caused in coverage.html oweing to multiple
"Plottable"s. One of these was extremely difficult to track down, so going with
the string style for now.

This reverts commit 59e620017d668d62311ce10e821edad49a20cd61.

Commit:
[Bar Plot tests] Further cleanup of variables.

crmorford committed:
give tests consistent naming

crmorford committed:
remove parentheses

crmorford committed:
make sure query point is always on screen

crmorford committed:
whitespace

Commit:
Change Pie Plot no-data entityNearest test to use width() and height()

crmorford committed:
Merge pull request #2945 from palantir/2925fix

entity nearest returns undefined instead of breaking, when no data in view

Commit:
Release version 1.16.2

jtlan committed:
Merge pull request #2947 from palantir/r1.16.2

Release version 1.16.2 (--> develop)

ztsai committed:
dragBoxLayer test refactor

softwords commented:
fixed fiddle for callback option 'explode' after change to new interface
http://jsfiddle.net/oaskqta0/28/

ztsai committed:
XDragBoxLayerTests refactor

ztsai committed:
refactor YDragBoxLayer test

ztsai committed:
guideLineLayerTests refactor

ztsai committed:
add hackhack comment for any casting

ztsai committed:
uninline const point declaration

ztsai committed:
uninline const point declaration

ztsai committed:
adjust expected position to depend on the orientation

ztsai committed:
Merge pull request #2951 from palantir/dragBoxLayerTests

dragBoxLayer test refactor

ztsai committed:
Merge pull request #2955 from palantir/guideLineLayerTests

guideLineLayerTests refactor

ztsai committed:
Merge pull request #2952 from palantir/xDragBoxLayerTests

XDragBoxLayerTests refactor

ztsai committed:
Merge pull request #2953 from palantir/yDragBoxLayerTests

refactor YDragBoxLayer test

Commit:
Sanitize newlines in domains

Previously, newlines would cause a failure, as the category axis
  uses SingleLineWrapper. Now, newlines are replaced with spaces.

crmorford committed:
tests

crmorford committed:
fix weight and padding

crmorford committed:
style

crmorford committed:
fix tests first

crmorford committed:
use isValidNumber, change positive to non-negative

Commit:
Support newlines instead of removing them

crmorford committed:
test non whole numbers

ztsai committed:
refactor selectionBoxLayerTests

TheBeruriahIncident committed:
Merge pull request #2959 from palantir/bugfix/handle-newlines-in-domains

Sanitize newlines in category scale domains

jtlan committed:
Merge pull request #2940 from palantir/2634refactor

[Testing Audit] Bar Plot

ztsai committed:
refactor DragLineLayerTests

ztsai committed:
group xScale tests and yScale tests

ztsai committed:
lint

ztsai committed:
Merge pull request #2965 from palantir/dragLineLayerTests

refactor DragLineLayerTests

ztsai committed:
Merge pull request #2962 from palantir/SelectionBoxLayerTEST

refactor selectionBoxLayerTests

jtlan committed:
Merge pull request #2914 from softwords/gitAttributes

Force git to recognize .d.ts files as text

Commit:
Merge branch 'develop' into interactionTestRefactor

crmorford committed:
use correct asserts, add NaN and number strings

Commit:
[Click/DoubleClick/Drag tests] Only remove svg if test succeeds.

Commit:
[Click/DoubleClick tests] Move test-class declarations before the tests

Commit:
[Click Interation tests] Adjust structure on callback registration tests

ztsai committed:
refactor gridlinesTests

Commit:
[Click Interaction tests] Restructure nesting of callback tests

Also make error messages generic to both mouse and touch cases.

Commit:
[Table tests] Move input-validation checks to their own describe()

Also made them a bit more "unit" by splitting the row and column tests
into separate it()s.

ztsai committed:
add tests for scales

crmorford committed:
Merge pull request #2960 from palantir/tableInput

Row/Column + Padding/Weight don't accept Infinity or negative numbers

ztsai committed:
remove test and fix typo

Commit:
[Click/DoubleClick tests] Switch to function-object for test callback.

softwords commented:
Hi @bluong you guys have clearly been deep in thought! Can you merge develop into objectConstancy at some point and I can keep my PR up to date with develop too.

Commit:
[DoubleClick Interaction tests] Restructure nesting of callback tests

Commit:
[Click/DoubleClick/Drag tests] Nest under describe("Interactions")

Commit:
[Click/DoubleClick/Drag tests] TSlint

Commit:
[Drag Interaction tests] Move registration/deregistration return type asserts

Commit:
[Click/DoubleClick tests] Fix linting issue with types

label-position rule in Travis rejects function object types with
the function signature as the first line

Commit:
[Drag Interaction tests] Restructure describe() nesting

ztsai committed:
Merge pull request #2967 from palantir/gridlinesTests

refactor gridlinesTests

Commit:
[Dispatcher tests] Clean up parent-level Dispatcher tests.

A more thorough cleaning (removing <any>s, etc) probably depends
on https://github.com/palantir/plottable/issues/2968.

Commit:
[Click/DoubleClick tests] Factor out constant points

Commit:
[Drag Interaction tests] Rename fake event triggering methods

Commit:
[Drag Interaction tests] Switch to function-object for test callbacks.

Commit:
[Drag Interaction tests] Redo nesting to split on registering/invoking

... instead of by event type.

Commit:
[Dispatcher] Change Dispatcher to use event-to-CallbackSet mapping.

Subclasses to follow.

Commit:
[Dispatcher Tests] Add InstrumentedDispatcher mock class to tests

Commit:
[Dispatchers] Convert Key, Mouse, and Touch to use new registration system

Commit:
[Dispatcher] Update tests and remove old callback registration system

Commit:
[Dispatchers] TSlint

ztsai committed:
refactor legendTests

Commit:
[Drag Interaction tests] Clean up touch-cancel test.

jtlan committed:
Merge pull request #2969 from palantir/2628refactor-dispatcher

[Dispatcher tests] Clean up parent-level Dispatcher tests.

Commit:
[Scale tests] Add one more level of nesting for Scale/QuantitativeScale

Commit:
[Scale tests] Split tests for Scale and QuantitativeScale

Commit:
[Scale tests] Nest tests under "onUpdate" describe()

Commit:
[QuantitativeScale tests] Clean up, skip nonvalid test

Test checks for https://github.com/palantir/plottable/issues/2336, but did so
in a way that didn't repro the bug. Not sure if we should fix it, since
processing strings may be out of scope for Scale logic.

Commit:
[Scale tests] Switch to a mock Scale subclass

Commit:
[Scale Tests] Use the power of the mock scale to add new tests!

jtlan committed:
Merge pull request #2893 from palantir/interactionTestRefactor

Issue #2628: refactor interaction unit tests

Commit:
[Scale tests] Add tests for autodomaining logic

ztsai committed:
grouping tests

ztsai committed:
refactor InterpolatedColorLegendTests

ztsai committed:
minor cleanup

ztsai committed:
refactor LabelTests

ztsai committed:
Merge pull request #2970 from palantir/LegendTests

refactor legendTests

Commit:
[QuantitativeScale tests] Add tests for TickGenerator behavior

Commit:
[QuantitativeScale tests] add tests for computing extent with bad values

Commit:
[QuantitativeScale tests] Collapse extent cases with forEach

Commit:
Revert "[QuantitativeScale tests] Collapse extent cases with forEach"

Semantics of the skipped test are better with the cases written out.

This reverts commit d3d6cb63bd67b7924da34dee7a9f187bd1e28b9a.

Commit:
[QuantitativeScale tests] Add test for rejecting negative padding values

Commit:
[Scale/QuantitativeScale tests] TSLint

ztsai committed:
Merge pull request #2972 from palantir/InterpolatedColorLegendTests

refactor InterpolatedColorLegendTests

ztsai committed:
restructure tests; remove usage of table

ztsai committed:
loose closeto_requirment to pass tests in linux

Commit:
[Scale/QuantitativeScale tests] Clean up a few tests for readability.

ztsai committed:
minor cleanup

ztsai committed:
Merge pull request #2973 from palantir/LabelTests

refactor LabelTests

jtlan committed:
Merge pull request #2974 from palantir/2640refactor-superclasses

[Testing Audit] Clean up Scale and QuantitativeScale tests

Commit:
[Drawer tests] Add new describe() containing tests for previously-untested methods

Will unify the describes later.

Commit:
[Drawer tests] Replace old test for totalDrawTime()

Commit:
[Drawer tests] Add asserts that check inputs to the projector

Commit:
[Drawer tests] Replace the test for selectionForIndex()

Commit:
[Drawer tests] Nest animation-timing tests under the main describe()

Commit:
[Drawer tests, Linear Scale tests] Get rid of top-level-polluting createMockDrawer method

Commit:
[Drawer tests] Combine drawing/selecting tests into a describe()

Commit:
[Drawer tests] Move "total draw time" test into the same describe() as the other animation-timing tests

Commit:
[Drawer tests] Add "has the correct CSS class" assert to drawing test

Commit:
[Drawer tests] Add HACKHACK to avoid #2777 (two-Plottables problem)

Commit:
[Label tests] fix implicit any.

Commit:
[Label tests] fix implicit any.

Commit:
[Line Drawer tests] Replace selectionForIndex() test

Commit:
[Line Drawer tests] Add test that it applies fill="none" style

Commit:
[Area Drawer tests] Add Area Drawer tests

Commit:
[Arc Drawer tests] Add tests for Drawers.Arc and Drawers.ArcOutline

jtlan committed:
Merge pull request #2977 from palantir/fixImplicitAny

[Label tests] fix implicit any.

ztsai committed:
refactor clusteredBarPlotTests

ztsai committed:
remove wheel plot

Commit:
[Drawer tests] Clean up animation-timing tests

Commit:
[ArcOutline Drawer tests] move to separate file

ztsai committed:
typo

jtlan committed:
Merge pull request #2978 from palantir/2622drawers

[Testing Audit] Added tests for Drawer subclasses.

jtlan committed:
Merge pull request #2976 from palantir/2622-parent-drawer

[Testing Audit] parent Drawer tests

ztsai committed:
consider overlay outside of component for pointerInteractions

Commit:
[Null Animator tests] Add tests for Animators.Null

ztsai committed:
Merge pull request #2985 from palantir/develop

merge develop into 2.0 branch

Commit:
[Easing Animator tests] Add getter/setter tests

Commit:
[Easing Animator tests] let -> const for existing tests

Commit:
[Easing Animator tests] Replace tests for total time computation

Replaced tests with independent checks of each setter's effects on time

Commit:
[Easing Animator tests] Remove misplaced totalTime() tests

Commit:
[Easing Animator tests] Factor out Animator initialization

ztsai committed:
Merge pull request #2979 from palantir/clusteredBarTests

refactor clusteredBarPlotTests

Commit:
[Easing Animator tests] Restructure maxTotalDuration() tests

Now tests the effects of increasing other parameters one at a time.

ztsai committed:
refactor pointerInteractionTests

Commit:
[Easing Animator tests] Use correct assert messages...

ztsai committed:
restructure describe block

jtlan committed:
Merge pull request #2987 from palantir/2622animators

[Testing Audit] Animators

ztsai committed:
Merge pull request #2989 from palantir/pointerInteraction

refactor pointerInteractionTests

ztsai committed:
Merge pull request #2982 from palantir/wheelPlot

remove wheel plot

ztsai committed:
refactor keyInteractionTests

ztsai committed:
Merge pull request #2991 from palantir/develop

merge develop into 2.0 branch

ztsai committed:
clean up touchDispatcherTests

softwords commented:
Hi @bluong noticing that you are working towards a version 2? If you have a moment to give me a quick heads-up on whether you are considering using any of this animation stuff? 
As time goes by I'm seeing conflicts inevitably appearing between these PRs and develop.... if you don't see yourselves using this PR then I won't try to keep up - I'll stay with my own fork which does what I need it to.

Commit:
Merge branch 'develop' into stage1AnimateJoin

# Conflicts:
#	test/drawers/drawerTests.ts

Commit:
post merge - revert drawerTests

ztsai committed:
moving out tests with mouse events

ztsai committed:
Merge pull request #2992 from palantir/touchDispatcher

clean up touchDispatcherTests

ztsai committed:
Merge pull request #2990 from palantir/keyInteractionTests

refactor keyInteractionTests

ztsai committed:
Merge pull request #2995 from palantir/develop

merge develop into 2.0 branch

ztsai committed:
get return types for Scale._getDomain() and Scale._getRange()

ztsai committed:
remove changes in subclasses

Commit:
Fix implicit-any in keyInteractionTests

Commit:
[Dispatchers] Rename _eventToCallback --> _eventToProcessingFunction

Old name was confusing because "callback" usually refers to callbacks
registered on the Dispatcher, not callbacks within the Dispatcher itself.

Commit:
[Key Interaction tests] Remove unused options argument in helper function

ztsai committed:
Merge pull request #2997 from palantir/scale

set return types for Scale._getDomain() and Scale._getRange()

jtlan committed:
Merge pull request #2999 from palantir/fixImplicitAny

Fix implicit-any in keyInteractionTests

Commit:
[Mouse Dispatcher] Use lastMousePosition() getter.

Commit:
Merge branch 'develop' into 2968refactor

Commit:
Upgrade to TS 1.7.3

Commit:
[TS1.7] Fix errors in DragBoxLayer.

Commit:
[TS1.7] Use this-typing on Animators.

Commit:
[TS1.7] Use this-typing on Axes.

Commit:
Merge branch 'develop' into develop-2.0

Commit:
[Touch Dispatcher tests] fix typing error

Commit:
Set _isTopLevelComponent based on whether the root node is an SVG

Previously was set to true and never reset!

Close #3004.

Commit:
[Dispatcher] rename variable for clarity

Commit:
Merge branch 'develop-2.0' into 2968refactor

jtlan committed:
Merge pull request #3000 from palantir/2968refactor

[Dispatchers] Refactor architecture for clarity

jtlan committed:
Merge pull request #3009 from palantir/3004fix

Set _isTopLevelComponent based on whether the root node is an SVG

ztsai committed:
[TS1.7] tslint dependency

ztsai committed:
[TS1.7] Use this-typing on Components

ztsai committed:
[TS1.7] Use this-typing on Dataset

ztsai committed:
[TS1.7] Use this-typing on Dispatchers

ztsai committed:
[TS1.7] Use this-typing on Drawer

ztsai committed:
[TS1.7] Use this-typing on Interaction

Commit:
Add backing <rect> to <svg> so Safari sees events.

ztsai committed:
[TS1.7] Use this-typing on Plots

ztsai committed:
[TS1.7] Use this-typing on Scales

Commit:
Switch to >= for backing size test

ztsai committed:
Merge pull request #2984 from palantir/interaction

pointerInteraction should not trigger under overlay

ztsai committed:
[TS1.7] more this-typing

ztsai committed:
remove deprecated methods and fields

ztsai committed:
Merge pull request #3010 from palantir/ts1.7upgrade

upgrade to ts 1.7.3

ztsai committed:
Merge pull request #3011 from palantir/deprecated

remove deprecated functions

Commit:
[Component tests] Add tests for safari event backing when root Component changes

Commit:
[Component tests] Add HACKHACKs for detach() calls made necessary by #3013.

jtlan committed:
Merge pull request #3012 from palantir/safariEvents

Add backing <rect> to catch events on Safari.

ztsai committed:
remove unused cached _computedWidth and _computeHeight

ztsai committed:
Merge pull request #3016 from palantir/remove-cache

remove unused cached variables

jtlan committed:
Merge pull request #3014 from palantir/develop-2.0

develop-2.0 --> develop

Commit:
Release version 2.0.0-rc1

jtlan committed:
Merge pull request #3017 from palantir/r2.0.0-rc1

Release version 2.0.0-rc1 (--> develop)

adidahiya committed:
EditorConfig, whitespace cleanup, README edits

adidahiya committed:
Update and alphabetize devDependencies

- Also update tsconfig.json to use 'outFile', which
  is available since TS 1.6

adidahiya committed:
Merge pull request #3025 from adidahiya/editorconfig-cleanup

EditorConfig, whitespace cleanup, README edits

adidahiya committed:
Merge pull request #3027 from adidahiya/update-deps

Update and alphabetize devDependencies

Commit:
Merge branch 'develop' into stage1AnimateJoin

# Conflicts:
#	plottable.d.ts
#	src/animators/nullAnimator.ts

Commit:
[QuantitativeScale] Compute padded domain based off of candidate domain, not current domain

Since scale() and invert() calls are used to compute the padded domain, the padded size
depends on the current domain and range, which causes problems if the domain or range
have been set to span zero width. This change temporarily sets the domain when computing
the padded domain.

Also took the opportunity to fix a test for snappingDomainEnabled to use less fragile numbers.

Close #3019.

Commit:
[Time Scale tests] let --> const in added test

jtlan committed:
Merge pull request #3029 from palantir/3019fix

[QuantitativeScale] Compute padded domain based off of candidate doma…

Commit:
Release version 2.0.0-rc2

jtlan committed:
Merge pull request #3030 from palantir/r2.0.0-rc2

Release version 2.0.0-rc2 (--> develop)

Commit:
Merge branch 'develop' into stage1AnimateJoin

Commit:
[StackedArea Plot] Default to no cropping by default

Cropped rendering on StackedArea is currently broken (as per #3032), so disabling it
by default for now until the algorithm can be fixed.

Commit:
[StackedArea Plot] Prevent croppedRenderingEnabled() from being set to true and log warning.

Due to #3032.

Commit:
[StackedArea Plot] Use correct warning.

jtlan committed:
Merge pull request #3033 from palantir/stackedAreaNoCrop

[StackedArea Plot] Default to no cropping by default

Commit:
Merge branch 'develop' into stage1AnimateJoin

Commit:
Release version 2.0.0

jtlan committed:
Merge pull request #3034 from palantir/r2.0.0

Release version 2.0.0 (--> develop)

Commit:
Merge branch 'develop' into stage1AnimateJoin

Commit:
[Plots.Scatter] Reset path attribute is now symbol with size 0 instead of empty string

This allows the Scatter Plot to animate corectly, with the symbols "growing"
to their final size.

Added a NoOpAnimator to the Mocks, which is used to blank the MAIN Animator
so we can inspect the reset state of a Plot.

Close #3037.

Commit:
[Plots.Scatter tests] Use afterEach() to clean up

Commit:
[Scatter Plot tests] Use d3 append-read to normalize path.

Repeated SVG commands appear to be optional in some browsers but not others (like IE).

jtlan committed:
Merge pull request #3039 from palantir/3037fix

[Plots.Scatter] "reset" path attribute is no longer empty string

Commit:
Merge branch 'develop' of https://github.com/palantir/plottable into stage1AnimateJoin

Commit:
fix noOpAnimator to Animator interface

adidahiya committed:
Enable additional TSLint rules

- no-conditional-assignment
- no-inferrable-types
- no-shadowed-variable

adidahiya committed:
Bump NPM dependencies

adidahiya committed:
Merge pull request #3049 from palantir/tslint-rules

Enable additional TSLint rules

adidahiya committed:
Floating grunt dependency

adidahiya committed:
Merge pull request #3054 from palantir/bump-deps

Bump NPM dependencies

Commit:
chai.d.ts

Commit:
Merge branch 'develop' into stage1AnimateJoin

m4dc4p committed:
ModifiedLogScale Documentation Update

Clarified usage of base, range and domain parameters.

adidahiya committed:
Merge pull request #3058 from m4dc4p/m4dc4p/3057-logScale-docs

ModifiedLogScale Documentation Update

Commit:
Remove installing typescript globally. Add solution for permission issues.

adidahiya committed:
Merge pull request #3062 from palantir/bugfix/update-development-guide

Update development guide on the README.md

m4dc4p committed:
Minor Refactor

* Make use of scale interface for setting/getting range.
* Use _range member consistently

endrjuskr committed:
Merge pull request #3064 from m4dc4p/m4dc4p/category-scale-fix

Category Scale - Minor Refactor

Commit:
Merge branch 'develop' into bugfix/bar-chart-modified-log-axis

Commit:
Get backing domain instead of adjusted domain.

adidahiya committed:
Merge pull request #3065 from palantir/bugfix/bar-chart-modified-log-axis

Bugfix/bar chart modified log axis

endrjuskr committed:
Nearest entity on plot group

Introduces PlotGroup component which exposes entityNearest on a collection of plots.

jimexist committed:
[PiePlot] Bug fix in pieplot where scale changes are not reflected upon dataset change

crmorford committed:
delete bad test, delete colors from dropdown, fix broken tests

crmorford committed:
Merge pull request #3076 from palantir/fixingQuicktests

delete bad test, delete colors from dropdown, fix broken tests

jkillian committed:
Prepare for NPM publishing (#3077)

* Bump lint deps

* Add proper dependencies, unbundle SVGTypewriter

* Prepare test npm release

adidahiya committed:
Fix typo in Gruntfile

adidahiya committed:
Release version 2.1.0

adidahiya committed:
Rename LinePlot#entityNearest to entityNearestByXThenY (#3079)

Fixes #3070.

LinePlot now uses the default euclidean distance algorithm for nearest entities.
The old X-then-Y algorithm is preserved in this newly renamed method.

adidahiya committed:
Release version 2.2.0

Commit:
Merge v2.20

# Conflicts:
#	plottable.d.ts
#	src/animators/animator.ts
#	test/drawers/drawerTests.ts
